src/pages/[gatewayUID]/details.tsx

```diff
@@ -1,32 +1,72 @@
 /* eslint-disable react/jsx-props-no-spreading */
-import type { GetServerSideProps, NextPage } from "next";
+import { useState, useEffect } from "react";
+import type { NextPage } from "next";
 import { useRouter } from "next/router";
 import { ParsedUrlQuery } from "querystring";
-import { useState } from "react";
-import { changeGatewayName } from "../../api-client/gateway";
+import { changeGatewayName, useGateway } from "../../api-client/gateway";
 import GatewayDetails from "../../components/elements/GatewayDetails";
 import { LoadingSpinner } from "../../components/layout/LoadingSpinner";
-import { services } from "../../services/ServiceLocator";
 import { getErrorMessage } from "../../constants/ui";
-import { ERROR_CODES } from "../../services/Errors";
-
+import { useNodes } from "../../api-client/node";
 import GatewayDetailViewModel from "../../models/GatewayDetailViewModel";
 import { getGatewayDetailsPresentation } from "../../components/presentation/gatewayDetails";
 
-type GatewayDetailsData = {
-  viewModel: GatewayDetailViewModel;
-  err?: string;
-};
+interface GatewayDetailsQueryInterface extends ParsedUrlQuery {
+  gatewayUID: string;
+}
 
-const GatewayDetailsPage: NextPage<GatewayDetailsData> = ({
-  viewModel,
-  err,
-}) => {
+const GatewayDetailsPage: NextPage = () => {
+  const [viewModel, setViewModel] = useState<GatewayDetailViewModel>({});
   const [isLoading, setIsLoading] = useState(false);
-  const router = useRouter();
-  const refreshData = async () => {
-    await router.replace(router.asPath);
-  };
+  const [err, setErr] = useState<string | undefined>(undefined);
+  const refetchInterval = 20000;
+
+  const { query } = useRouter();
+  const { gatewayUID } = query as GatewayDetailsQueryInterface;
+
+  const {
+    isLoading: gatewayLoading,
+    error: gatewayErr,
+    data: gateway,
+    refetch: gatewayRefetch,
+  } = useGateway(gatewayUID, refetchInterval);
+
+  const {
+    isLoading: nodesLoading,
+    error: nodeErr,
+    data: nodes,
+  } = useNodes([gatewayUID]);
+
+  useEffect(() => {
+    if (gatewayErr) {
+      setErr(getErrorMessage(gatewayErr.message));
+    }
+    if (nodeErr) {
+      setErr(getErrorMessage(nodeErr.message));
+    }
+    if (!nodeErr && !gatewayErr) {
+      setErr(undefined);
+    }
+  }, [gatewayErr, nodeErr]);
+
+  useEffect(() => {
+    if (gatewayLoading) {
+      setIsLoading(true);
+    }
+    if (nodesLoading) {
+      setIsLoading(true);
+    }
+    if (!gatewayLoading && !nodesLoading) {
+      setIsLoading(false);
+    }
+  }, [gatewayLoading, nodesLoading]);
+
+  useEffect(() => {
+    if (gateway && nodes) {
+      const gatewayViewModel = getGatewayDetailsPresentation(gateway, nodes);
+      setViewModel(gatewayViewModel);
+    }
+  }, [gateway, nodes]);
```

Comment:
Is there an easier way we could do eventually?
 
Seems somewhat verbose.

I'd probably feel a lot better if all the `use` logic were pulled out into `<GatewayDetailsLiveLoader>` (or something like that) and this `details.tsx` were strictly for NextJS-related code.

Reply:
Very possibly - for final implementation, lifting it into another component would probably work.

It's just somewhat tricky because both useQuery calls must be resolved before this `presentation` function can run properly